src/bin/cql-stress-cassandra-stress/java_generate/distribution/mod.rs

```diff
@@ -0,0 +1,32 @@
+pub trait Distribution: Send + Sync {
+    fn next_i64(&mut self) -> i64;
+    fn next_f64(&mut self) -> f64;
+    fn set_seed(&mut self, seed: i64);
+}
```

Comment:
The distinction is introduced for efficiency (and correctness) purposes. It's needed for the distributions that use `java_random::Random` under the hood (which is not atomic). 

For these distributions, we will implement `Distribution` in a way, that each implementation will have its own copy of `java_random::Random` (as a member in a struct).

As of `AtomicDistribution`, me and @avelanarius decided to go with the `thread_local` approach (which is much faster than wrapping with mutex). In this case, there will be a `thread_local` copy of `java_random::Random` used to implement `next_{i/f}64_atomic` methods.

Note that distributions can be shared between that threads, as well as used by a single tokio task. We can't always use `AtomicDistributions` - because of the `thread_local`s. Assume one tokio task running on thread X changes the state of the underlying RNG (e.g. `set_seed(0)`), then it needs to `await` something - we switch to other task on the same thread X. Then the second task calls `set_seed(1)`. It needs to `await` something. We switch back to the first thread - it expects to `seed == 0`, which is obviously not the case. We strongly rely on the deterministic nature of c-s value generation logic.

When it comes to the distributions that will be shared across all tokio tasks, their state cannot be manipulated with `set_seed` method. What is also important, is the fact that there will be only one instance of such distribution during program execution.

 Please, correct me if the scenario I provided can never happen. From what I understood, `tokio` uses a thread pool - which implies that multiple tokio tasks can run on the same thread.

Reply:
What I understood both before and after you wrote the above is that `AtomicDistribution` is suitable for multithreaded use (by avoiding data races, which is by the way obligatory in _safe Rust_ to implement `Send + Sync`), but offers limited functionality (no `set_seed()`) in order to prevent race conditions, which could lead to logical errors (by an interlaced execution that you described).

The question I had in mind was: can't we make `Distribution` delegate to `AtomicDistribution`'s methods for generating random numbers? It's invalid though, because `AtomicDistribution` requires `Send + Sync` and `Distribution` does not.

Many thanks for clarification. I think it may be a good idea to put the above consideration in a comment, placed next to `Distribution` and `AtomicDistribution` trait definition.

Reply:
I added a comprehensive explanation of why do we need to introduce such distinction. It's located in the comments, next  to `Distribution` and `AtomicDistribution`. It also explains how are both types of distributions used in the row generation logic.